dask_expr/expr.py

```diff
@@ -679,12 +681,29 @@ def _task(self, index: int):
         -------
         task: tuple
         """
-        args = tuple(self._blockwise_arg(op, index) for op in self.operands)
-        if self._kwargs:
-            return (apply, self.operation, args, self._kwargs)
+        args, kwargs = self._get_args_kwargs_from_operands(
+            lambda x: self._blockwise_arg(x, index)
+        )
+        if self._kwargs or kwargs:
+            return (
+                apply,
+                self.operation,
+                (tuple, list(args)),
+                {**self._kwargs, **kwargs},
+            )
         else:
             return (self.operation,) + args
 
+    def _get_args_kwargs_from_operands(self, op_func) -> tuple[tuple, dict]:
```

Comment:
I don't think there is anything wrong with the proposal here, but I do worry about "readability".  It *may* be a little easier to interpret if we just define `_args` and `_kwargs` properties, and use them throughout `Blockwise` (in lieu of `self.operands`. What do you think of something like this?:


```python
class Blockwise(Expr):
    ...
    _keyword_only = None

    @functools.cached_property
    def _meta(self):
        return self.operation(
            *[arg._meta if isinstance(arg, Expr) else arg for arg in self._args],
            **self._kwargs,
        )

    @functools.cached_property
    def _kwargs(self) -> dict:
        if self._keyword_only:
            return {
                p: self.operand(p)
                for p in self._parameters
                if p in self._keyword_only
            }
        return {}

    @functools.cached_property
    def _args(self) -> list:
        if self._keyword_only:
            args = [
                self.operand(p)
                for p in self._parameters
                if p not in self._keyword_only
            ] + self.operands[len(self._parameters):]
            return args
        return self.operands

    ....

    def _task(self, index: int):
        args = [self._blockwise_arg(op, index) for op in self._args]
        if self._kwargs:
            return (apply, self.operation, args, self._kwargs)
        else:
            return (self.operation, *args)
```

Reply:
We can do this as well, but we have to apply ``_blockwise_arg`` to the kwargs values as well, which loses some of the advantages

Reply:
Hmm. It makes me a bit nervous that  you are saying Kwargs may hold other expressions now.

Reply:
Can you share an example where this is the case (sorry if it’s right in front of me)?

Reply:
pandas deprecated a bunch of positional arguments in 2.0, ``clip`` is one example where this could be problematic (there are probably more)

Reply:
Hmm. I’m not sure I understand the clip example. Is this because upper or lower can be another expression?

Reply:
Yes exactly, that's a possibility.

Reply:
Ah, okay. I understand where you are coming from then.

Keep in mind that you cannot include a key in the kwargs argument to `apply`. You can only include a keys within the `args`. For example, this assertion will fail:

```python
import dask
from dask.utils import apply

def func(x, y=None):
    return y


dsk = {
    "key": (apply, func, ["key-a"], {"y": ("key-b", 0)}),
    ("key-a", 0): "expr-a",
    ("key-b", 0): "expr-b",
}

assert dask.get(dsk, "key") == "expr-b"
```
Therefore, I don't think there is any good reason to apply `_blockwise_arg` to the kwargs. If we want to allow users to include another collection/expression in the kwargs, we will need to introduce a new version of `apply` (which I think we can probably punt on here).

Reply:
I wasn't aware of this. In that case we can avoid the function for kwargs. Agreed, let's punt on this here